src/tests/intg/files_ops.py

```diff
@@ -0,0 +1,155 @@
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+import os
+import ent
+import os.path
+import tempfile
+import pytest
+from util import backup_envvar_file, restore_envvar_file
+
+@pytest.fixture
+def passwd_ops_setup(request):
+    pwd_file = os.environ["NSS_WRAPPER_PASSWD"]
+    backup_envvar_file("NSS_WRAPPER_PASSWD")
+    request.addfinalizer(lambda: restore_envvar_file("NSS_WRAPPER_PASSWD"))
+    pwd_ops = PasswdOps(pwd_file)
+    return pwd_ops
+
+
+@pytest.fixture
+def group_ops_setup(request):
+    grp_file = os.environ["NSS_WRAPPER_GROUP"]
+    backup_envvar_file("NSS_WRAPPER_GROUP")
+    request.addfinalizer(lambda: restore_envvar_file("NSS_WRAPPER_GROUP"))
+    grp_ops = GroupOps(grp_file)
+    return grp_ops
+
+
+@pytest.fixture
+def group_db_setup(request):
+    group = request.param
+    grp_ops = group_ops_setup(request)
+    grp_ops.groupadd(**group)
+    ent.assert_group_by_name(group['name'], group)
+    return grp_ops
+
+
+class FilesOps(object):
+    """
+    A naive implementation of operations as a basis for user or group
+    operations. Uses rename to (hopefully) trigger the same fs-level
+    notifications as shadow-utils would.
+    """
+    def __init__(self, file_name):
+        self.file_name = file_name
+        self.tmp_dir = os.path.dirname(self.file_name)
+
+    def _get_named_line(self, name, contents):
+        for num, line in enumerate(contents, 0):
+            pname = line.split(':')[0]
+            if name == pname:
+                return num
+        raise KeyError("%s not found" % name)
+
+    def _read_contents(self):
+        with open(self.file_name, "r") as pfile:
+            contents = pfile.readlines()
+        return contents
+
+    def _write_contents(self, contents):
+        tmp_file = tempfile.NamedTemporaryFile(dir=self.tmp_dir, delete=False)
+        tmp_file.writelines(contents)
+        tmp_file.flush()
+
+        os.rename(tmp_file.name, self.file_name)
+
+    def _append_line(self, new_line):
+        contents = self._read_contents()
+        contents.extend(new_line)
+        self._write_contents(contents)
+
+    def _subst_line(self, key, line):
+        contents = self._read_contents()
+        kindex = self._get_named_line(key, contents)
+        contents[kindex] = line
+        self._write_contents(contents)
+
+    def _del_line(self, key):
+        contents = self._read_contents()
+        kindex = self._get_named_line(key, contents)
+        contents.pop(kindex)
+        self._write_contents(contents)
+
+        contents = self._read_contents()
+
+
+class PasswdOps(FilesOps):
+    """
+    A naive implementation of user operations
+    """
+    def __init__(self, file_name):
+        super(PasswdOps, self).__init__(file_name)
+
+    def _pwd2line(self, name, uid, gid, passwd, gecos, dir, shell):
```

Comment:
Try to use different name then `dir`. It is a built-in function in python and you override it. It could cause issues when debugging. The same name is used also on different places.